File: WeatherForecast.mjs
//code by Chanel Yang. 2024.10.25
//References to HTML elements
const locationElement = document.getElementById('location');
const temperatureElement = document.getElementById('temperature');
const weatherDescElement = document.getElementById('weatherDesc');
const warningElement = document.getElementById('warning');
const weatherDisplay = document.getElementById('weatherDisplay');
const loadingElement = document.getElementById('loadingWeather');

//function to fetch weather data from python backend
async function fetchWeatherFromBackend(lat, lon) {
    console.log("Calling fetch Weather from back end");
    const url = 'http://localhost:5000/location';//backend URL
    
    try {
        const response = await fetch(url, {
            method: 'POST',//the request uses the POST HTTP method
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({latitude: lat, longitude: lon})
        }); //initiate a fetch request to the URL
        const data = await response.json(); //convert the HTTP file received to JSON file
        console.log("response received from backend", data);
        if (response.ok) { //if the response is received
            displayWeather(data);

            if (data.warning && data.warning.length > 0) { 
                console.log("got warning")//if severe weather alerts present
                displayWarning(data.warning);
            }else {
                warningElement.innerText = 'No severe weather warnings.'; //display no warnings
            }
        } else {
            loadingElement.innerText = 'Fail to receive weather information.';
        }
    } catch (error) {
        loadingElement.innerText = 'Error fetching weather data. Please report to the developers.';
    }
}

//function to display the weather data on frontend client side
function displayWeather(weatherData) {
    console.log("Calling displayWeather");
    console.log(weatherData);
    loadingElement.style.display = 'none'; //the loading message is hidden
    weatherDisplay.style.display = 'block'; //display the weather data

    //display weather details
    locationElement.innerText = `Location:  ${weatherData.name}, ${weatherData.sys.country}`;
    temperatureElement.innerText = `Temperature: ${weatherData.main.temp} °C`;
    weatherDescElement.innerText = `Conditions: ${weatherData.weather[0].description}`;
    
}

//function to display sever weather warnings
function displayWarning(warning) {
    console.log("calling get warning");
    warningElement.innerText = warning;
}

//main function. get location and fetch data
function getLocation() {
    console.log("Calling get location");

    if (navigator.geolocation) {//check if geolocation is supported
        console.log("accessed geolocation");
        navigator.geolocation.getCurrentPosition(position => {
            const lat = position.coords.latitude;//get the latitude
            const lon = position.coords.longitude;//get the longitude

            fetchWeatherFromBackend(lat, lon);
        }, () => {
            loadingElement,innerText = 'Unable to access your location.';
        });
    } else {
        loadingElement.innerText = 'Geolocation is not supported.';
    }
}

//call getlocation function on page load
getLocation();

File: WeatherForecast.py
from flask import Flask, request, jsonify
import requests
from plyer import notification
app = Flask(__name__)
API_key = "8a691e21287b4229f68d9f12641eb5e3"

def get_weather_data(latitude, longitude):
    url = f"http://api.openweathermap.org/data/2.5/weather?lat={latitude}&lon={longitude}&appid={API_key}&units=metric"
    response = requests.get(url)
    data = response.json()
    if data["cod"] != "404":
        return data
    else:
        return None

def Check_extreme_weather(data):
    weather_condition = data["weather"][0]["description"]
    temp = data["main"]["temp"]
    extreme_weather = False
    alert_message = "!!!Extreme Weather Alert!!!"
    alert_message_width = len(alert_message)
    if "thunderstorm" in weather_condition.lower() or "heavy rain" in weather_condition.lower():
        alert_message_weather = f"SEVERE WEATHER: {weather_condition}."
        final_alert_message = f"{alert_message}\n{alert_message_weather.center(alert_message_width)}"
        extreme_weather = True
    if temp > 37 or temp < -5:
        alert_message_temp = f"EXTREME TEMPERATURE: {temp}°C"
        final_alert_message = f"{alert_message}\n{alert_message_temp.center(alert_message_width)}"
        extreme_weather = True
    if extreme_weather:
        return final_alert_message
    else:
        return "No extreme weather detected."
    
def Location():
    data = request.get_json()
    latitude = data.get("latitude")
    longitude = data.get("longitude")
    weather_data = get_weather_data(latitude, longitude)
    if weather_data:
        alert_message = Check_extreme_weather(weather_data)
        return jsonify({"message": alert_message})
    else:
        return jsonify({"message": "Weather data not found:(."}), 404
    
if __name__ == "__main__":
    app.run(debug=True)